src/api.rs

```diff
@@ -519,6 +374,163 @@ async fn api_error_adapter(response: reqwest::Response) -> Result<reqwest::Respo
+        .ok_or(InputError::InvalidFileName {
+            path: file_path.clone(),
+        })?;
+
+    // Request upload URL
+    let url = format!("{}/v2/upload/presigned-url", base_url_with_params);
+    let params = [("api_key", api_key.clone())];
+    let url = reqwest::Url::parse_with_params(&url, &params)
+        .map_err(|error| ApiError::InvalidParameters { error })?;
+
+    let request_body = UploadRequest {
+        filename: file_name.to_string(),
+    };
+    let upload_url_response = client.post(url).json(&request_body).send().await?;
+    let upload_url_response = api_error_adapter(upload_url_response)
+        .await?
+        .json::<UploadUrlResponse>()
+        .await
+        .map_err(|error| ApiError::DeserializationFailure { error })?;
+
+    // Progress stuff
+    let file_total_size = file.metadata().await?.len();
+    let mut file_reader = ReaderStream::new(file);
+    let mut multi_progress: Option<MultiProgress> = if !no_progress_bar {
+        Some(MultiProgress::new())
+    } else {
+        None
+    };
+    let file_progress_bar;
+    let file_body;
+    if !no_progress_bar {
+        let sty = ProgressStyle::with_template(
+            "{spinner:.blue} [{elapsed_precise}] [{wide_bar:.cyan/blue}] {bytes}/{total_bytes} ({bytes_per_sec}, {eta})"
+        )
+        .unwrap()
+        .progress_chars("#>-");
+
+        let pb = ProgressBar::new(file_total_size);
+        pb.enable_steady_tick(Duration::from_millis(80));
+        file_progress_bar = multi_progress.as_mut().unwrap().add(pb);
+        file_progress_bar.set_style(sty.clone());
+        let mut file_progress = 0u64;
+        let file_stream = async_stream::stream! {
+            while let Some(chunk) = file_reader.next().await {
+                let file_progress_bar = file_progress_bar.clone();
+                if let Ok(chunk) = &chunk {
+                    let new = min(file_progress + (chunk.len() as u64), file_total_size);
+                    file_progress = new;
+                    file_progress_bar.set_position(new);
+                    if file_progress >= file_total_size {
+                        file_progress_bar.finish_and_clear();
+                    }
+                }
+                yield chunk;
+            }
+        };
+        file_body = Body::wrap_stream(file_stream);
+    } else {
+        file_body = Body::wrap_stream(file_reader);
+    }
+
+    let s3_response = client
+        .put(upload_url_response.url.clone())
+        .header("Content-Length", file_total_size)
+        .body(file_body)
+        .send()
+        .await?;
+    api_error_adapter(s3_response).await?;
+
+    Ok(upload_url_response.file_path.clone())
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct UploadRequest {
+    filename: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct UploadUrlResponse {
+    file_path: String,
+    url: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct CreateRunRequest {
```

Comment:
`rename` will be added in the next PR.
Also, there are some field names mistakes that will be fixed in the next PR too.